### PR TITLE
KAFKA-7391 Introduce close(Duration) to Producer and AdminClient inst…

### PR DESCRIPTION
**Pull request by chia7712**
KIP-367: https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=89070496

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Review by viktorsomogyi — COMMENTED** [review]
LGTM overall, although I'm a bit unsure in removing the close(long, TimeUnit) from KafkaAdminClient. Probably it is fine since users shouldn't use that directly + it's evolving, but a committer will tell.

**Comment by chia7712** [comment]
@viktorsomogyi  Thanks for reviews!!

> although I'm a bit unsure in removing the close(long, TimeUnit) from KafkaAdminClient. 

This PR put default implementation of close(long, TimeUnit) to KafkaAdminClient's parent class (AdminClient.java) so it should be fine to remove close(long, TimeUnit) from KafkaAdminClient.

**Comment by omkreddy** [comment]
@chia7712  Thanks for the Patch. Can you rebase the PR and add @SuppressWarnings annotations for deprecations?

**Comment by chia7712** [comment]
>  Can you rebase the PR and add @SuppressWarnings annotations for deprecations?

done.

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java
Comment (omkreddy) [inline clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:1136]: Do we need to add a negative timeout check like we did in [consumer](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java#L2101)?

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java
Comment (omkreddy) [inline clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:444]: Do we need to add a negative timeout check like we did in [consumer](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java#L2101)?

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java
Comment (chia7712) [reply clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:1136]: add the check and related test (see KafkaProducerTest.testNegativeTimeout)

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java
Comment (chia7712) [reply clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:444]: add the check and related test (see KafkaAdminClientTest.testNegativeTimeout)

**Comment by omkreddy** [comment]
retest this please

**Comment by chia7712** [comment]
> retest this please

done

**Comment by omkreddy** [comment]
@chia7712 By commenting "retest this please",  we can re-trigger the jenkins build.

**Comment by chia7712** [comment]
Sorry for my misunderstanding :(

**Comment by omkreddy** [comment]
not sure why tests are failing/not completing. retriggering the tests again.
retest this please

**Comment by omkreddy** [comment]
retest this please

**Comment by omkreddy** [comment]
retest this please

**Comment by omkreddy** [comment]
retest this please

**Comment by omkreddy** [comment]
retest this please

**Comment by omkreddy** [comment]
@chia7712 some reason tests are failing due to out of memory. Can you verify if we are missing anything?

**Comment by chia7712** [comment]
> Can you verify if we are missing anything?

Copy that

**Comment by chia7712** [comment]
Run all tests for this PR locally. It fails sometimes because of OOM but it pass mostly.
Will dig into OOM later...

File: clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java]: Can we update to 2.2?

File: clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java]: nit: mentioning the time unit seems redundant.

File: clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java:91]: I wonder if a safer way to do this from a compatibility perspective would be to provide a default method for `close(Duration)` which invokes `close(long, TimeUnit)`. Similarly for the producer.

**Review by hachikuji — COMMENTED** [review]
Sorry for the review delay. Left a few small comments.

File: clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java
Comment (chia7712) [reply clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java:91]: Got it. However, it will be hard to remove deprecated methods if the compatibility guarantee includes the "inheritance"...

Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java:91]: I think deprecation covers this. Eventually at a major release version, we can drop the old methods. But it's reasonable to expect that we don't break compatibility until the deprecation window has closed.

Comment (chia7712) [reply clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java:91]: > I think deprecation covers this. Eventually at a major release version, we can drop the old methods. But it's reasonable to expect that we don't break compatibility until the deprecation window has closed.

Thanks for the explanation!

**Comment by hachikuji** [comment]
retest this please

**Comment by hachikuji** [comment]
@chia7712 The patch looks good, but we seem to be having trouble getting a clean build. Can you take a look at the build failures?

**Comment by chia7712** [comment]
> The patch looks good, but we seem to be having trouble getting a clean build. Can you take a look at the build failures?

Copy that

**Comment by chia7712** [comment]
@hachikuji Please take a look at this PR. Thanks!

File: clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java]: I asked @ijuma for a second opinion on this approach. It was my suggestion, but it seemed annoying that extensions are forced to implement the deprecated API. He brought up the good point that we have to be able to break extension compatibility when we add new admin APIs anyway. So maybe I had this wrong and we should add the default method to the other `close`. What do you think?

Comment (chia7712) [reply clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java]: > e brought up the good point that we have to be able to break extension compatibility when we add new admin APIs anyway. So maybe I had this wrong and we should add the default method to the other close. What do you think?

Agreed. We have to break extension compatibility in order to push kafka forward. I will update PR later.

**Comment by chia7712** [comment]
@hachikuji Any suggestions?

**Review by hachikuji — APPROVED** [review]
LGTM. Thanks for the patch!

**Pull request merged**
